kips/kip-2.md

```diff
@@ -0,0 +1,462 @@
+        }
+        // remove last item
+        events[numEvents - 1] = address(0);
+        numEvents--;
+      }
+    }
+    
+    events.length = numEvents;
+
+    if (_newEvent != address(0)) {
+      events[numEvents] = _newEvent;
+      events.length++;
+    }
+  }  
+}
+```
+
+**IEvent, Event**
+
+The event contracts (i.e. the existing `Conference` contract). Just the rough minimum requirements are given here.
+
+```solidity
+/**
+ * Interface for events
+ */
+interface IEvent {
+  function hasEnded() external view returns (bool);
+  function getPayout(address _addr) external view returns (uint256);
+  function getDeposit(address _addr) external view returns (uint256);
+}
+
+
+
+contract Event is IEvent {
+  // ... existing fields
+  IUserPot public userPot;
+
+  constructor (
+    address _userPot,
+    string _name,
+    uint256 _deposit,
+    uint256 _limitOfParticipants,
+    address _owner
+  ) public {
+    userPot = IUserPot(_userPot);
+    // ...
+  }
+
+  function register() external payable onlyActive{
+    require(registered < limitOfParticipants, 'participant limit reached');
+    require(!isRegistered(msg.sender), 'already registered');
+
+    // send to user pot
+    userPot.deposit.value(msg.value)(msg.sender);
+
+    registered++;
+    participantsIndex[registered] = msg.sender;
+    participants[msg.sender] = Participant(registered, msg.sender);
+
+    emit RegisterEvent(msg.sender, registered);
+  }
+
+
+  function getPayout(address _user) external view returns (uint256) {
+    if (!ended || !isAttended(_user)) {
+        return 0;
+    }
+    return payoutAmount;
+  }
+
+  function getDeposit(address _user) external view returns (uint256) {
+    return deposit;
+  }
+
+  function hasEnded() external view returns (bool) {
+    return ended;
+  }
+
+  // .... rest of the code almost same as the existing Conference contract
+}
+```
+
+
+## Future considerations
+
+The following topics has been discussed with respect to this KIP.
+
+### Automatic withdrawal
+
+Event though this change eliminates the need to call `withdraw()` per event, the user still has to withdraw manually at some point if they want to have their ETH back.
+
+Some users expressed that they would prefer that their ETH gets sent back automatically. This could be achieved in future by having the `withdraw()` function be called on behalf of a user by a third-party. But realistically speaking, users are probably better off not _needing_ to withdraw anymore, and only withdrawing later on when they actually _want_ to.
+
+### Support for different payout strategies
+
+There is some interest in different payout strategies, e.g. using sweepstake and for different logic when handling a paid event. This KIP will enable us to cater for such variations as it detaches event-related logic from the money management side of things.
+
+### ERC20 token support
+
+The code above was written for ETH-funded events. Some minor modifications will be needed for ERC-20 support.
```

Comment:
How does this work. We create `ERC20UserPotImpl` and `ERC20Event` to abstract the logic like we do right now?

Reply:
No, better idea to incorporate it into the existing Pot. You basically need to track the current balance of the user for every possible token.

Reply:
Inside the pot, the code which calculates the storage key string for the balance should take into account the token being dealt with.